lang/rust/Makefile

```diff
@@ -0,0 +1,182 @@
+RUST_INSTALL_TARGET_FILENAME:=$(PKG_NAME)-$(PKG_VERSION)-$(RUSTC_TARGET_ARCH)-install.tar.xz
+
+# Rust-lang has an uninstall script
+RUST_UNINSTALL:=$(CARGO_HOME)/lib/rustlib/uninstall.sh
+
+# This checks to see if the dist installation already exists
+IS_RUSTC_INSTALLED:=$(or $(and $(wildcard $(CARGO_HOME)/bin/rustc),true),false)
+IS_CARGO_INSTALLED:=$(or $(and $(wildcard $(CARGO_HOME)/bin/cargo),true),false)
+
+# MUSL Target Flags
+MUSL_CONFIGURE_ARGS = \
+  --set=target.$(RUSTC_TARGET_ARCH).ar=$(TARGET_AR) \
+  --set=target.$(RUSTC_TARGET_ARCH).cc=$(TARGET_CC_NOCACHE) \
+  --set=target.$(RUSTC_TARGET_ARCH).cxx=$(TARGET_CXX_NOCACHE) \
+  --set=target.$(RUSTC_TARGET_ARCH).linker=$(TARGET_CC_NOCACHE) \
+  --set=target.$(RUSTC_TARGET_ARCH).musl-root=$(TOOLCHAIN_DIR) \
+  --set=target.$(RUSTC_TARGET_ARCH).ranlib=$(TARGET_RANLIB)
+
+# CARGO_HOME is an environmental
+HOST_CONFIGURE_OPTS += CARGO_HOME="$(CARGO_HOME)"
+
+# The Flags are Common between HOST and TARGET
+RUST_COMMON_ARGS := \
+  --build=$(RUSTC_HOST_ARCH) \
+  --dist-compression-formats=xz \
+  --enable-llvm-link-shared \
+  --enable-llvm-plugins \
+  --enable-missing-tools \
+  --enable-ninja \
+  --disable-sanitizers \
+  --release-channel=stable \
+  --enable-cargo-native-static \
+  ${MUSL_CONFIGURE_ARGS}
+
+##
+# Because OpenWrt uses default configure flags that are not recognized
+# we override the default and set our own.
+#
+# Rust HOST Configuration Arguments
+HOST_CONFIGURE_ARGS = \
+  --target=$(RUSTC_TARGET_ARCH),$(RUSTC_HOST_ARCH) \
+  --host=$(RUSTC_HOST_ARCH) \
+  --prefix=$(CARGO_HOME) \
+  --bindir=$(CARGO_HOME)/bin \
+  --libdir=$(CARGO_HOME)/lib \
+  --sysconfdir=$(CARGO_HOME)/etc \
+  --datadir=$(CARGO_HOME)/share \
+  --mandir=$(CARGO_HOME)/man \
+  ${RUST_COMMON_ARGS}
+
+# TODO: For future use
+# Rust TARGET Configuration Args
+#CONFIGURE_ARGS = \
+#  --target=$(RUSTC_TARGET_ARCH) \
+#  --host=$(RUSTC_TARGET_ARCH) \
+#  --prefix=$(CONFIGURE_PREFIX) \
+#  --bindir=$(CONFIGURE_PREFIX)/bin \
+#  --libdir=$(CONFIGURE_PREFIX)/lib \
+#  --sysconfdir=$(CONFIGURE_PREFIX)/etc \
+#  --datadir=$(CONFIGURE_PREFIX)/share \
+#  --mandir=$(CONFIGURE_PREFIX)/man \
+#  $(RUST_COMMOM_ARGS) \
+#  --set=llvm.cxxflags='$(RUST_LLVM_CXXFLAGS)'
+
+# Add rust_config.mk Debug flags
+#include ./rust_config.mk
+
+define Host/Prepare
+	# Ensure rust temp directory
+	[ -d $(RUST_TMP_DIR) ] || \
+	  mkdir -p $(RUST_TMP_DIR)
+
+	$(call Host/Prepare/Default)
+endef
+
+define Host/Uninstall
+	# Call the Uninstall script
+	[ -f $(RUST_UNINSTALL) ] && \
+	  $(BASH) $(RUST_UNINSTALL) || echo No Uninstall
+
+	rm -rf $(RUST_TMP_DIR)
+endef
+
+# Makes and then packages the dist artifacts
+define Host/Compile
+	cd $(HOST_BUILD_DIR) && \
+	  $(PYTHON) x.py --config ./config.toml dist build-manifest cargo llvm-tools \
+	    rustc rust-std rust-src
+
+	$(call Host/PackageDist)
+endef
+
+# Distribution Artifacts are packaged by Host/PackageDist
+# We just need to extract and install
+define Host/Install
```

Comment:
I do not know which is the best practice in this regard. surely it can be done that way.

Reply:
meson was and python3 or go are examples.

Since this lives in the package feed repository, you can just put any .mk helper files in the root of this package and other packages can include the .mk file based on their own location.

See 468572af01e708aa017d1c58ff53070d1a255077 (where it still lived here before moving it to the base repo).
Or `git grep python3-package.mk` or `git grep golang-package.mk`.

Reply:
so it would be `$(STAGING_DIR_HOSTPKG)/lib/rust/build.mk`

Reply:
I assume the approach from python/README.md is deprecated?

Reply:
If you mean the `$(TOPDIR)` part, I, and I guess all others , prefer relative paths, but both would work I guess :)

Reply:
The meson approach copies the helper files in $(STAGING_DIR_HOSTPKG), the python approach makes you refer to the file in feeds, I'd go the meson way.

Right now I'm having absurd problems with the choice of bundling a (stale) libc but not its headers nor a compiler aware of it so I guess this issue is secondary...

Reply:
meson copies itself to staging, because it's just a bunch of python scripts. It's similar to a toolchain that installs itself there. I assume your rust host package does the same.

But then there're .mk helper files, and by that I mean pure OpenWrt specific files to help creating OpenWrt runtime packages. I assumed we're talking about that, in which case the feed location is appropriate.

Reply:
I see, I misread how meson worked based on what @maximilien-baumann suggested.